tests/unit/EntityId/PrefixingEntityIdParserTest.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Tests\EntityId;
+
+use Wikibase\DataModel\Entity\BasicEntityIdParser;
+use Wikibase\DataModel\Entity\EntityIdParsingException;
+use Wikibase\DataModel\Entity\ItemId;
+use Wikibase\DataModel\Entity\PropertyId;
+use Wikibase\DataModel\Services\EntityId\PrefixingEntityIdParser;
+
+/**
+ * @covers Wikibase\DataModel\Services\EntityId\SuffixEntityIdParser
+ *
+ * @license GPL-2.0+
+ */
+class PrefixingEntityIdParserTest extends \PHPUnit_Framework_TestCase {
+
+	/**
+	 * @dataProvider validInputProvider
+	 */
+	public function testParse( $prefix, $input, $expected ) {
+		$parser = new PrefixingEntityIdParser( $prefix, new BasicEntityIdParser() );
+		$this->assertEquals( $expected, $parser->parse( $input ) );
+	}
+
+	public function validInputProvider() {
+		return array(
+			array( 'wikidata:', 'Q1337', new ItemId( 'wikidata:Q1337' ) ),
+			array( 'foo:', 'P123', new PropertyId( 'foo:P123' ) ),
+			array( 'foo:', 'bar:Q1337', new ItemId( 'foo:bar:Q1337' ) ),
+		);
+	}
+
+	public function testParse_prefixIncluded() {
+		$parser = new PrefixingEntityIdParser( 'wikidata:', new BasicEntityIdParser() );
+		$this->assertEquals( new ItemId( 'wikidata:Q1337' ), $parser->parse( 'wikidata:Q1337' ) );
+	}
+
+	/**
+	 * @dataProvider invalidInputProvider
+	 */
+	public function testParse_invalidInput( $prefix, $input ) {
+		$parser = new PrefixingEntityIdParser( $prefix, new BasicEntityIdParser() );
+		$this->setExpectedException( EntityIdParsingException::class );
+		$parser->parse( $input );
```

Comment:
I don't think this type of dataProvider usage pays for itself. How about having two methods:

* testGivenInvalidId_exceptionIsThrown
* testGivenEmptyId_exceptionIsThrown

And now I wrote this I realized that actually you are testing the decorated parser, BasicEntityIdParser, which is real production code. The one valid thing being tested here is that exceptions thrown by the decorated parser are not somehow caught. You only need one test case for that. And you can easily avoid binding to BasicEntityIdParser by using a test double.